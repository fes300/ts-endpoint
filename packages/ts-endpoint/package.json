{
  "name": "ts-endpoint",
  "version": "2.0.1-alpha.0",
  "description": "Simple endpoint definition for fullstack safety.",
  "main": "index.js",
  "repository": "https://github.com/fes300/ts-endpoint/tree/master/packages/ts-endpoint",
  "bugs": {
    "url": "https://github.com/fes300/ts-endpoint/issues?q=is%3Aopen+is%3Aissue+project%3Afes300%2Fts-endpoint%2F3"
  },
  "author": "Federico Sordillo <federicosordillo@gmail.com>",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "npm-publish-beta": "if [ -f lib/package.json ] ; then npm publish ./lib --tag beta; else echo echo \"no need to publish\"; fi",
    "npm-publish": "if [ -f lib/package.json ] ; then npm publish ./lib; else echo echo \"no need to publish\"; fi",
    "postversion": "cp package.json lib && cp -r lib lib2 && mv lib2 lib/lib",
    "build": "rm -rf ./lib && tsc",
    "type-test": "jest --config=jest.typetests-config.js",
    "runtime-test": "jest --config=jest.config.js",
    "test": "pnpm type-test && pnpm runtime-test"
  },
  "devDependencies": {
    "io-ts": "^2.2.20"
  },
  "dependencies": {
    "fp-ts": "^2.13.1",
    "ts-io-error": "workspace:*",
    "typelevel-ts": "^0.4.0"
  }
}
