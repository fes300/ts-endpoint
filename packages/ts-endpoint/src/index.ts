export * from './Endpoint';
export * from './helpers';
export * from './ResourceEndpoint';

