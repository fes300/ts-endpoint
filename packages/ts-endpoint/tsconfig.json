{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "lib": ["esnext"],
    "rootDir": "./src",
    "outDir": "./lib",
    "tsBuildInfoFile": "lib/.tsbuildinfo",
    "esModuleInterop": true
  },
  "include": ["src"]
}
