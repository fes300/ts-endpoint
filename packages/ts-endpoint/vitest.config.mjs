import { defineProject, mergeConfig } from 'vitest/config';
import vitestBaseConfig from '../../vitest.config.mjs';

export default mergeConfig(
  vitestBaseConfig,
  defineProject({
    test: {
      name: 'ts-endpoint',
      typecheck: { enabled: true },
    },
  })
);
