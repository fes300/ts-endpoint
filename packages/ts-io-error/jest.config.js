const base = require('../../jest.config.base.js');

module.exports = {
  ...base,
  displayName: 'ts-io-error',
};
