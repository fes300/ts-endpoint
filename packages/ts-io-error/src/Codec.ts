// -------------------------------------------------------------------------------------
// Codec
// -------------------------------------------------------------------------------------

export interface Codec<A, B = A, R = never> {
  Type: A;
  Encoded: B;
  Context: R;
}

// type AnyOrRuntime<N> = [N] extends [never]
//   ? any
//   : N extends Record<any, Codec<any, any>>
//   ? { [k in keyof N]: runtimeType<N[k]> }
//   : never;

export type UndefinedsToPartial<R> = R extends Record<string, any>
  ? { [K in keyof R as undefined extends R[K] ? K : never]?: R[K] } & {
      [K in keyof R as undefined extends R[K] ? never : K]: R[K];
    }
  : R;

// serialized
export type serializedType<C> = C extends Codec<any, infer A> ? A : never;
export type RecordSerialized<N> = N extends Record<string, Codec<any>>
  ? { [K in keyof N]: serializedType<N[K]> }
  : never;

export type RecordCodecSerialized<N> = N extends RecordCodec<infer A> ? RecordSerialized<A> : never;

export type SerializedType<N> = N extends RecordCodec<infer A>
  ? RecordCodecSerialized<A>
  : N extends Codec<any, any>
  ? serializedType<N>
  : N extends Record<string, Codec<any>>
  ? RecordSerialized<N>
  : unknown;

// encoded
export type runtimeType<C> = C extends Codec<infer B, any> ? B : never;

export type RecordEncoded<N> = N extends Record<string, Codec<any>>
  ? { [K in keyof N]: runtimeType<N[K]> }
  : never;

export type EncodedType<N> = N extends Codec<any, any>
  ? runtimeType<N>
  : N extends RecordCodec<infer A>
  ? RecordEncoded<A>
  : N extends Record<string, Codec<any>>
  ? RecordEncoded<N>
  : unknown;

export type PartialRecordEncoded<N> = N extends Record<string, Codec<any>>
  ? UndefinedsToPartial<RecordEncoded<N>> | undefined
  : never;

export interface RecordCodec<A extends Record<string, Codec<any>>>
  extends Codec<RecordSerialized<A>, RecordEncoded<A>> {
  fields: A;
}

export type RecordCodecEncoded<N> = N extends RecordCodec<infer B> ? RecordEncoded<B> : never;

export type PartialRuntimeType<N> = N extends Codec<any, any>
  ? UndefinedsToPartial<runtimeType<N>>
  : never;

export type PartialSerializedType<N> = N extends Codec<any, any>
  ? UndefinedsToPartial<serializedType<N>>
  : never;
