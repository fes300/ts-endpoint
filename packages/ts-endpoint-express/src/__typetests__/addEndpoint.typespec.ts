import * as express from 'express';
import { left, right } from 'fp-ts/Either';
import * as O from 'fp-ts/Option';
import * as t from 'io-ts';
import { Endpoint } from 'ts-endpoint/lib';
import { IOError } from 'ts-io-error/lib';
import { RecordCodec } from 'ts-io-error/lib/Codec';
import { buildIOError, GetEndpointSubscriber } from '../index';

const getEndpoint = Endpoint({
  Input: {
    Params: t.type({ id: t.string }),
    Headers: t.type({ auth: t.string }),
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const postEndpoint = Endpoint({
  Input: {
    Body: t.type({ content: t.string }),
  },
  Method: 'POST',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
});

const postEndpointWithErrors = Endpoint({
  Input: {
    Body: t.type({ content: t.string }),
  },
  Method: 'POST',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Errors: { 404: t.type({ error: t.string }), 401: t.type({ baz: t.string }) },
});

const router = express.Router();
const registerRouter = GetEndpointSubscriber(buildIOError);
const AddEndpoint = registerRouter(router);

// @dts-jest:fail:snap should not allow empty calls
AddEndpoint();
// @dts-jest:fail:snap won't compile if output of controller is wrong
AddEndpoint(getEndpoint, ({ params: { id } }) => () => {
  console.log(id);
  return Promise.resolve(right({ body: { crayons: [22] }, statusCode: 200 }));
});

// @dts-jest:fail:snap won't compile if trying to access non existent param
AddEndpoint(getEndpoint, ({ params: { id, bar } }) => () => {
  console.log(bar, id);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 200 }));
});

// @dts-jest:fail:snap won't compile if trying to access non defined body
AddEndpoint(getEndpoint, ({ params: { id }, body: { foo } }) => () => {
  console.log(id, foo);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 200 }));
});

type OutputOrUndefined<T> = T extends RecordCodec<any, any, infer R> ? R : undefined;
const v = t.type({ u: t.string });
type a = OutputOrUndefined<typeof v>;
const s: RecordCodec<any, any> = v;

// @dts-jest:pass:snap correct constructions should work
AddEndpoint(getEndpoint, ({ params: { id } }) => () => {
  console.log(id);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 200 }));
});

// @dts-jest:pass:snap correct constructions should work
AddEndpoint(postEndpoint, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(right({ body: { crayons: ['brown'] }, statusCode: 201 }));
});

// @dts-jest:fail:snap you cannot return a badly formatted error
AddEndpoint(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(left({ foo: 'baz' }));
});

AddEndpoint(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  // @dts-jest:fail:snap you cannot return a badly formatted error
  return Promise.resolve(left(new IOError('error', { kind: 'KnownError', error: 'foo' })));
});

AddEndpoint(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(
    // @dts-jest:fail:snap you cannot return a badly formatted error
    left(new IOError('error', { kind: 'KnownError', status: 401, body: { error: 'foo' } }))
  );
});

AddEndpoint(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);
  return Promise.resolve(
    // @dts-jest:pass:snap you can return a well formatted error
    left(new IOError('error', { kind: 'KnownError', status: 401, body: { baz: 'foo' } }))
  );
});

export const buildMaybeError = (errors: unknown[]) => {
  return O.some({
    kind: 'DecodingError',
    errors,
  });
};

declare module '../HKT' {
  interface URItoKind<A> {
    Option: O.Option<A>;
  }
}

const maybeRouter = express.Router();
const registerMaybeRouter = GetEndpointSubscriber<'Option'>(buildMaybeError);
const AddMaybeEndpoint = registerMaybeRouter(maybeRouter);

// @dts-jest:fail:snap you cannot return a badly formatted error with a different error builder
AddMaybeEndpoint(postEndpointWithErrors, ({ body: { content } }) => () => {
  console.log(content);

  return Promise.resolve(left(O.some({ kind: 'KnownError', status: 401, body: { error: 'foo' } })));
});
