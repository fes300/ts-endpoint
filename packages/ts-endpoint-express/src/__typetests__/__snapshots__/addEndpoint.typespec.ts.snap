// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct constructions should work (type) should match snapshot 1`] = `"void"`;

exports[`correct constructions should work (type) should match snapshot 2`] = `"void"`;

exports[`should not allow empty calls (type) should match snapshot 1`] = `"Expected 2 arguments, but got 0."`;

exports[`won't compile if output of controller is wrong (type) should match snapshot 1`] = `
"Type '() => Promise<Either<IOError<never>, { body: { crayons: number[]; }; statusCode: number; }>>' is not assignable to type 'TaskEither<IOError<never>, HTTPResponse<{ crayons: string[]; }>>'.
  Type 'Promise<Either<IOError<never>, { body: { crayons: number[]; }; statusCode: number; }>>' is not assignable to type 'Promise<Either<IOError<never>, HTTPResponse<{ crayons: string[]; }>>>'.
    Type 'Either<IOError<never>, { body: { crayons: number[]; }; statusCode: number; }>' is not assignable to type 'Either<IOError<never>, HTTPResponse<{ crayons: string[]; }>>'.
      Type 'Right<{ body: { crayons: number[]; }; statusCode: number; }>' is not assignable to type 'Either<IOError<never>, HTTPResponse<{ crayons: string[]; }>>'.
        Type 'Right<{ body: { crayons: number[]; }; statusCode: number; }>' is not assignable to type 'Right<HTTPResponse<{ crayons: string[]; }>>'.
          Type '{ body: { crayons: number[]; }; statusCode: number; }' is not assignable to type 'HTTPResponse<{ crayons: string[]; }>'.
            The types of 'body.crayons' are incompatible between these types.
              Type 'number[]' is not assignable to type 'string[]'.
                Type 'number' is not assignable to type 'string'."
`;

exports[`won't compile if trying to access non defined body (type) should match snapshot 1`] = `"Property 'foo' does not exist on type 'undefined'."`;

exports[`won't compile if trying to access non existent param (type) should match snapshot 1`] = `"Property 'bar' does not exist on type '{ id: string; }'."`;

exports[`you can return a well formatted error (type) should match snapshot 1`] = `"<E = never, A = never>(e: E) => Either<E, A>"`;

exports[`you cannot return a badly formatted error (type) should match snapshot 1`] = `
"Type '() => Promise<Either<{ foo: string; }, HTTPResponse<{ crayons: string[]; }>>>' is not assignable to type 'TaskEither<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
  Type 'Promise<Either<{ foo: string; }, HTTPResponse<{ crayons: string[]; }>>>' is not assignable to type 'Promise<Either<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>>'.
    Type 'Either<{ foo: string; }, HTTPResponse<{ crayons: string[]; }>>' is not assignable to type 'Either<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
      Type 'Left<{ foo: string; }>' is not assignable to type 'Either<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
        Type 'Left<{ foo: string; }>' is not assignable to type 'Left<IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>>'.
          Type '{ foo: string; }' is missing the following properties from type 'IOError<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>': details, status, name, message"
`;

exports[`you cannot return a badly formatted error (type) should match snapshot 2`] = `"Object literal may only specify known properties, and 'error' does not exist in type '{ kind: "KnownError"; status: 404; body: { error: string; }; } | { kind: "KnownError"; status: 401; body: { baz: string; }; }'."`;

exports[`you cannot return a badly formatted error (type) should match snapshot 3`] = `"Object literal may only specify known properties, and 'error' does not exist in type '{ baz: string; }'."`;

exports[`you cannot return a badly formatted error with a different error builder (type) should match snapshot 1`] = `
"Type '() => Promise<Either<O.Option<{ kind: string; status: number; body: { error: string; }; }>, HTTPResponse<{ crayons: string[]; }>>>' is not assignable to type 'TaskEither<Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
  Type 'Promise<Either<Option<{ kind: string; status: number; body: { error: string; }; }>, HTTPResponse<{ crayons: string[]; }>>>' is not assignable to type 'Promise<Either<Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>>'.
    Type 'Either<Option<{ kind: string; status: number; body: { error: string; }; }>, HTTPResponse<{ crayons: string[]; }>>' is not assignable to type 'Either<Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
      Type 'Left<Option<{ kind: string; status: number; body: { error: string; }; }>>' is not assignable to type 'Either<Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>, HTTPResponse<{ crayons: string[]; }>>'.
        Type 'Left<Option<{ kind: string; status: number; body: { error: string; }; }>>' is not assignable to type 'Left<Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>>'.
          Type 'Option<{ kind: string; status: number; body: { error: string; }; }>' is not assignable to type 'Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>'.
            Type 'Some<{ kind: string; status: number; body: { error: string; }; }>' is not assignable to type 'Option<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>'.
              Type 'Some<{ kind: string; status: number; body: { error: string; }; }>' is not assignable to type 'Some<{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }>'.
                Type '{ kind: string; status: number; body: { error: string; }; }' is missing the following properties from type '{ 404: TypeC<{ error: StringC; }>; 401: TypeC<{ baz: StringC; }>; }': 404, 401"
`;
